torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+import torch
+from torch.fx import Node
+from torch.fx._compatibility import compatibility
+from torch._subclasses.fake_tensor import FakeTensorMode, FakeTensor
+from torch.utils._pytree import tree_map
+from torch.multiprocessing.reductions import StorageWeakRef
+
+
+
+import _operator
+from enum import Enum
+import itertools
+from typing import Optional, Set
+from collections import defaultdict
+
+__all__ = ['reinplace']
+
+class _ViewType(Enum):
+    NonView = 0
+    SingleOutputView = 1
+    MultiOutputView = 2
+
+def _is_view_op(tgt):
+    if tgt is not None and isinstance(tgt, torch._ops.OpOverload):
+        schema = tgt._schema
+        if len(schema.arguments) > 0:
+            first_arg = schema.arguments[0]
+            # check if op is a view
+            return first_arg.alias_info is not None and not first_arg.alias_info.is_write
+
+def _get_view_type(tgt) -> _ViewType:
+    if tgt is not None and isinstance(tgt, torch._ops.OpOverload):
+        schema = tgt._schema
+        if len(schema.arguments) > 0:
+            first_arg = schema.arguments[0]
+            # check if op is a view
+            if first_arg.alias_info is not None and not first_arg.alias_info.is_write:
+                # check if op is a multi-output view
+                if '*' in first_arg.alias_info.after_set:
+                    return _ViewType.MultiOutputView
+                else:
+                    return _ViewType.SingleOutputView
+    return _ViewType.NonView
+
+
+# Stores a bunch of metadata related to functionalization each node.
+# Relevant metadata:
+# n.meta['fake_result']: FakeTensor (same type as the output of the node, but with FakeTenors instead of Tensors)
+#   The fake tensor output from running the current node
+# n.meta['node_usages']: List[List[int, Node]]
+#   A list of all nodes that take in the current node as an input
+#   Each element is actually an (int, Node) pair, specifying the index of the node in the graph.
+# n.meta['view_of']: Node
+#   If the current node n is a view of some base tensor, the 'view_of' field tells us which
+#   view node was used to generate the current node (a view tensor).
+#   This information actually makes `fake_result` redundant, but we can use `fake_result`
+#   to sanity check that our aliasing information is correct.
```

Comment:
Hmmmmm... technically, should `fake_result` also have recorded view metadata that also records this? (Not just strides, but the view function chain that we use to reconstruct views.) I admit it is probably not in a form that is easily accessible from Python.

Reply:
Is that the autograd view metadata? Or something else?

Reply:
View metadata

Reply:
Hmm if this is the `DifferentiableViewMeta` in autograd, then don't we need to worry about inference mode again? (I guess we could just temporarily disable it internally while tracing).

Or are you suggesting to try to re-use the `ViewMeta` chains constructed from functionalization? hmm.


Either way I might leave this as a followup :)